### ISSUE TITLE
One resource can be managed by multiple stacks

### ISSUE THREAD

**Issue by loekd** [Azure/deployment-stacks#15]
**Describe the bug**

Due to an [issue](https://github.com/Azure/deployment-stacks/pull/14) in the tutorial, I created multiple stacks that managed the same set of resources. (in this case the 4 storage accounts)

I would expect this to be impossible, as different stacks can have different behaviors to deal with resource lifecycle.

**To Reproduce**
1. follow the tutorial as it [currently](https://github.com/Azure/deployment-stacks/commit/ef0616ea7a5f5393e8a1437c0c17623697c422f8) is.
2. Do not purge the resources.
3. You will have 2 stacks governing the same resources.

```
Get-AzSubscriptionDeploymentStack

Id                : /subscriptions/mysub/providers/Microsoft.Resources/deploymentStacks/mySubStack
Name              : mySubStack
ProvisioningState : succeeded
UpdateBehavior    : purgeResources
Location          : westeurope
CreationTime(UTC) : 22-10-2021 06:35:05
ManagedResources  : /subscriptions/mysub/resourceGroups/dvstrldtstrg1/providers/Microsoft.Storage/storageAccounts/dvstrldtstzbl4bnkm4vvlgb
                    /subscriptions/mysub/resourceGroups/dvstrldtstrg2/providers/Microsoft.Storage/storageAccounts/dvstrldtstzbl4bnkm4vvlgc
                    /subscriptions/mysub/resourceGroups/dvstrldtstrg2/providers/Microsoft.Storage/storageAccounts/dvstrldtstzbl4bnkm4vvlgd
DeploymentId      : /subscriptions/mysub/providers/Microsoft.Resources/deployments/mySubStack-2021-10-22-06-59-58-10f8b
SnapshotId        : /subscriptions/mysub/providers/Microsoft.Resources/deploymentStacks/mySubStack/snapshots/2021-10-22-06-59-58-10f8b

Id                : /subscriptions/mysub/providers/Microsoft.Resources/deploymentStacks/stack
Name              : stack
ProvisioningState : succeeded
UpdateBehavior    : purgeResources
Location          : westeurope
CreationTime(UTC) : 22-10-2021 06:45:26
ManagedResources  : /subscriptions/mysub/resourceGroups/dvstrldtstrg1/providers/Microsoft.Storage/storageAccounts/dvstrldtstzbl4bnkm4vvlgb
                    /subscriptions/mysub/resourceGroups/dvstrldtstrg2/providers/Microsoft.Storage/storageAccounts/dvstrldtstzbl4bnkm4vvlgc
                    /subscriptions/mysub/resourceGroups/dvstrldtstrg2/providers/Microsoft.Storage/storageAccounts/dvstrldtstzbl4bnkm4vvlgd
DeploymentId      : /subscriptions/mysub/providers/Microsoft.Resources/deployments/stack-2021-10-22-06-50-18-88bdf
SnapshotId        : /subscriptions/mysub/providers/Microsoft.Resources/deploymentStacks/stack/snapshots/2021-10-22-06-50-18-88bdf
```

**Expected behavior**
I would expect to see an error when creating 2 stacks that govern the same resource.

**Comment by bmoore-msft** [comment]
This is intentional as scanning all Azure resources across subs (and in some cases tenants) is not a problem we can reasonably solve in a performant manner with the current design of ARM.  Once locking is available you can prevent this by locking a resource.

@mumian we should doc this...

**Comment by miqm** [comment]
@bmoore-msft - what about keeping with each resource  metadata with deployment stack id? This shouldn't be hard and would only to check resources touched by the template. Of course there will be some problem with nested deployments, but it wouldn't require scanning entire tenant.

**Comment by slavizh** [comment]
@miqm that would require a change in every resource provider, including extensions and proxy resources probably. My guess this would be very hard and time consuming change to do across all Azure resources.

**Comment by miqm** [comment]
Isn’t the Resource Manager responsible for sending requests to RPs? The validation of stack assignment could occur before RM does the calls.

**Comment by slavizh** [comment]
@miqm what RP has as properties and what are those properties are not controlled by Resource Manager. For example - https://docs.microsoft.com/en-us/rest/api/hybridcompute/machines/create-or-update#systemdata system data. It is something that needs to be implemented by every RP thus you use appearing RP by RP and version by version.

**Comment by miqm** [comment]
Not sure I follow. When you define a template, you specify resource and properties, so you can easily manage stack assignments. You don't need to know which are available, you just need to know which are defined during current stack update and which one were there in previous updates.

**Comment by bmoore-msft** [comment]
@miqm - I see what you're after and this is an interesting idea...  if I follow...

_When a stack creates a resource it can "tag" it with some metadata that indicates the managing stack resource id.  Now, the next time any stack looks at deploying a resource, it can check to see if that tag is present and behave accordingly._

Today, there isn't any way for a resource (deploymentStacks) to add some global metadata to any other resource.  Keep in mind that deploymentStacks (and the deployment engine) are a separate service from ARM where the metadata would be stored.  Could we do it?  Sure (we own the code ;)), but it's an uphill battle as there's a pretty high bar for something like this - we'd want to make sure we've vetted other solutions as well.  TLDR; this is a pretty hard problem to solve...  For example, I'm not sure what concurrency/race conditions might still exist given that ARM is declarative.  I suspect there are some RPs that would handle overlapping PUTS just fine, so you could still end up with a resource in multiple stacks.

That said let's pivot the discussion...  

What problem(s) are created by putting resources into multiple stacks? 

If these pain points are severe enough (or problems unsolvable) that gives us more feedback on how important it is to go after the problem across ARM.

**Comment by henrybeen** [comment]
I might be wrong, but wouldn't this tag be what I propose in #44 , including a scenario?

**Comment by miqm** [comment]
> What problem(s) are created by putting resources into multiple stacks?

I think the biggest problem is that you can remove resource by updating a completely different stack. Locks are not a solution - they should (IMHO) prevent you from removing critical resources. But in some cases you still would like to manage the resource in a particular stack (i.e. queues/topics in a service bus) and be able to remove them when removed from stack. playing with locks would add too much complexity.

I'd really would like to see deployment stacks to work like terraform state but a managed one, with RBAC and all that resource-related stuff.

> Today, there isn't any way for a resource (deploymentStacks) to add some global metadata to any other resource.

Can this resource-deployment-stack-metadata be kept in the deployment stack RP? it would be a simple (big, but simple) key-value set. A simple check to DeploymentStack RP by AzureRM before making a call should do the job.

On the other hand, I'd like to have ability to say that although a particular resource's lifecycle is managed by some stack, I'd like to include it in my template (data type in terraform) so I can update some properties. Unless if we're talking about lifecycle it's not just exists - does not exist but also keeping all defined properties and their values in sync with the data saved in stack.

> Today, there isn't any way for a resource (deploymentStacks) to add some global metadata to any other resource.

What about extension resources? Wouldn't this be some kind of a global metadata?

**Comment by bmoore-msft** [comment]
@henrybeen - I think #15 and #44 are very similar...

@miqm - going to keep this part focused on the "problem statement"...

>I think the biggest problem is that you can remove resource by updating a completely different stack. Locks are not a solution - they should (IMHO) prevent you from removing critical resources. But in some cases you still would like to manage the resource in a particular stack (i.e. queues/topics in a service bus) and be able to remove them when removed from stack. playing with locks would add too much complexity.

It sounds like the scenario is that an adjacent/unrelated operation could inadvertently remove a resource.  So resource is in both stack A and B and stack B owner decides it's not needed anymore and does a purge.  Someone who *thinks* they own the resource deletes it not realizing it's being used.

This actually is the scenario that locks/deny are intended for... the same thing could happen independent of stack B.  Someone sees a resource in a RG, doesn't think it's being used and deletes it.  Membership in a stack cannot prevent deleting, but locks/deny can.

>I'd really would like to see deployment stacks to work like terraform state but a managed one, with RBAC and all that resource-related stuff.

Can you expand on this?  I'm not terribly familiar with TF State but afaiki they don't have the capability to prevent "accidents" through TF state...

**Comment by miqm** [comment]
> Membership in a stack cannot prevent deleting, but locks/deny can.

Yes, I do agree with this. I still want to have ability for stack A to delete a resource once removed from stack, but if I need to slightly adjust that resource from deployment to stack B or just by mistake in bicep I’ll forget putting `existing` in the resource, and then I remove resource (or add existing) - I don’t want stack B to remove it.  
so in case when I deploy a stack and it will try to manage a resource lifecycle that is already managed by a different stack - I want to have some option to deal with. Most common action  would be to explicitly say that this resource, although is in the template, it’s lifecycle shouldn’t be managed here (force detach of it even if stack has purge). A resource should have only one master :)

As for terraform state - now when you deploy hcl code, terraform creates it’s state and needs to keep it somewhere. In Azure it’s usually a blob storage. But basically it can be any file capable storage. you can limit access to the state file by setting blob contributor role on container where state file is kept. Same thing I’d expect from stacks - to limit who can manipulate the state.
However you would still need at least a contributor role on resource group or subscription to deploy resources anyway, so separate role for stacks might be an overkill and do not have any value. This could be discussed separately.

**Comment by bmoore-msft** [comment]
TBH, I think the proper model is that you can't put the same resource in stack A and B at the same time and if it were simple we'd probably do it...  in reality this probably happens POST GA if it happens.  If so then we probably have a means to do #44 but the scenarios for that still needs some refining.

re: TF - the "recommended" model for a stack would be to deploy it at the parent scope... so a RG template should be in a stack that exists at sub scope.  That allows separation of roles for managing the resources - i.e. I can lock them and only users who have access to stacks at sub scope can change the stack/resources...  similar (not identical) to RBAC - separate the roles.

**Comment by miqm** [comment]
I agree with you that one resource should belong to one stack only. Hopefully you will find a way to do this.

One more case for you to consider when planning this feature - i.e. we might have a Service Bus managed in one stack (infrastructure) but queues and topics managed by different (application) stacks. Same thing might be with sql server and and databases or cdn profile and endpoints...

**Comment by slavizh** [comment]
As we are discussing cases: What about if stack A has the resource as detach but stack B has it as purge. For me this should be allowed. Overall I get the problem but this should be end user choice if resource can be in two stacks. Probably there should be some design guidance in stacks documentation to not recommend that scenario. For me the bigger problem with resource being in more than one stack is that different properties can be applied to that resource so that resource could always be changing properties depending from which stack you are deploying it. Deleting a resource is important operation for lifecycle but it is not the number 1 operation like deploying and changing resource properties. In any case I would opt in for some warning message that says you have this resource in Stack A as well. May be could be part of What-If functionality or similar to that pre-flight/validation check.

**Comment by bmoore-msft** [comment]
@miqm - my "default demo" for stacks does something similar with parent child (sqlServer & serverFarm in one stack, db and site in another)...  Are you seeing any issues in these cases?  (or how do you envision seeing it affect "the plan"?) 

@slavizh - the properties delta you touched on is a good one - that would mean in a drift detection (+remediation) scenario stack A shows a diff (so I fix it) then stack b shows the diff (then I fix it) and repeat...  I get that as a platform feature we could allow that, though that doesn't feel like the ```pit of success``` so would want to make sure we understand the scenarios for allowing it... though unfortunately out of the gate I don't think we can prevent it so may be stuck there.

**Comment by loekd** [comment]
I believe this is a valuable feature, that helps users avoid running into problems with conflicting desired state configurations, accidental deletes, and un-auditable infrastructure (because which one is the truth?). 

Having a stack create a deny assignment, locking out every principal except the creator, would require the use of a different principal for every deployment stack. Locks would get in the way when managing just a single stack (unlock, deploy, lock, over and over again).

Storing a (hidden) map of 'deployment stack id' vs. resourceID - designating ownership - at subscription could work.

**Comment by bmoore-msft** [comment]
> Having a stack create a deny assignment, locking out every principal except the creator, would require the use of a different principal for every deployment stack. Locks would get in the way when managing just a single stack (unlock, deploy, lock, over and over again).

Good point...  very good point.

**Comment by matsest** [comment]
I think introducing a strict locking for one resource -> one stack could create some additional issues around the management of these "locks" that ultimately can be avoided with more well thoughout template design, RBAC (and in edge cases Policy).

From my experience with Terraform and management of Terraform state in shared environments, most of this can be avoided with clear scoping of the state (i.e. which resources are included in the stack and which are not) + following good RBAC practices. If stack owner A and B both have e.g. Owner permissions, (traditional) resource locks can be bypassed anyway..

However, I think the idea of some metadata that a resource is part of a stack could be helpful and e.g. give a warning to users trying to manage the resource with others stacks or in other ways (portal, ps, etc.) I'm sure it's not trivial to expose such a warning across all possible interfaces (deploymentstacks, cli's, arm/bicep, portal, api calls..), but at least within the DeploymentStacks user experience would be a good starting point.

@bmoore-msft: I don't know if this is more related to the ARM engine, or within the scope of DeploymentStacks, but is an "ignore" argument on resource properties something you've considered ([from tf](https://www.terraform.io/language/meta-arguments/lifecycle#ignore_changes))? From experience with "shared states" this is (depending on rp design) sometimes a bit troublesome - like e.g. the vnet/subnet issue (can't redeploy vnet withouth subnets property set) or other properties that with is intentionally not set in initial deploy template (e.g. network acl on key vaults).

**Comment by bmoore-msft** [comment]
re: "ignore" - that feels like it's similar to PATCH which we have talked about but PATCH is more intentional (IMO) than ignore.  Generally, we try to avoid the "fixes" that cover up the warts and try to go after the warts - even though that's a bit slower than we'd like.  TF can't do that, they have to respond how they can...

re:

>like e.g. the vnet/subnet issue (can't redeploy vnet withouth subnets property set) or other properties that with is intentionally not set in initial deploy template (e.g. network acl on key vaults).

I get the vnet/subnet issue - not sure I follow on the network acls - can you expand on that (i.e. what's the problem scenario?)

**Comment by matsest** [comment]
> re: "ignore" - that feels like it's similar to PATCH which we have talked about but PATCH is more intentional (IMO) than ignore. Generally, we try to avoid the "fixes" that cover up the warts and try to go after the warts - even though that's a bit slower than we'd like. TF can't do that, they have to respond how they can...

I agree with the reasoning, and I'm aware it's an ongoing effort. As long as ARM resource providers are well designed to handle declarative configurations and make correct API calls it shouldn't really be an issue. The vnet/subnet is perhaps an edge case.. But great effort from you + the Bicep team to improve the underlying causes in the RP's! :muscle: 

> I get the vnet/subnet issue - not sure I follow on the network acls - can you expand on that (i.e. what's the problem scenario?)

My bad, I was thinking that the stack was overwriting unset properties in "managed" resources - but I see that it behaves like a regular ARM deploy.. So it's not a problem scenario with DeploymentStacks really (I need to translate some larger more complex deploys to stacks to get a better feel of this).. 

Not setting the property in the template is the correct way to go to ignore changes for that property if it's changed outside deployment..

**Comment by bmoore-msft** [comment]
Update:

Our current thinking on this behavior is to rely on read-only denyAssignments to prevent having a resource being added to multiple stacks.  So essentially when you add ```denyWrite``` to the resource via the "original" stack, no subsequent stack would be able to add/update the resource.

The approach is certainly logical - we're looking for feedback to see if there are any downsides to this approach.

**Comment by bmoore-msft** [comment]
I wanted to add a bit more to the proposal above regarding making resources "read-only" in order to prevent the resource being added to multiple stacks.  This does mean that excludedActions and/or excludedPrincipals would likely be needed in a number of scenarios to avoid blocking things like VM Shutdown or retrieving account keys (think any POST method for a resource).  If that adds more to think about with the approach...

**Comment by slavizh** [comment]
I do not have anything against additional options.

**Comment by bmoore-msft** [comment]
@slavizh - can you expand?  I'm not sure I follow...

**Comment by slavizh** [comment]
@bmoore-msft I mean I am fine with the additional proposal.

**Comment by bmoore-msft** [comment]
Any concerns about the POST actions/operations?  I.e. in order to get a storageAccount key, something would need to be exluded from the denyAssignment (principal and/or action)...  Not sure how laborious that's going to be, even though technically possible.

**Comment by slavizh** [comment]
There are plenty of examples where you will need to do some operations and those operations shouldn't be stopped. Sometimes you even have to generate SaS token instead of using the primary key for storage account. A lot of services that have autoscale mechanisms which change properties based on the scale configuration, etc. So I would say may be in some cases you will have to go even on property level which as far as I know it is not available. Of course without having the functionality available I do not know if these features will be completely blocked or somehow they have their own way of working without being blocked by deny assignments.

**Comment by mumian** [comment]
Add @timwarner-msft for awareness.

**Comment by bmoore-msft** [comment]
> There are plenty of examples where you will need to do some operations and those operations shouldn't be stopped. Sometimes you even have to generate SaS token instead of using the primary key for storage account. 

Ack - this is my main concern.  By using the "denyWrite" option you'd prevent these things by default.  We have to (do it by default) otherwise it's not possible to truly enable such a scenario where you don't want it done.  The owner of the stack can exclude those actions, the denyAssignment api supports that, it's just that you have to know, at the time of stack creation, that you need to exclude `Microsoft.Storage/storageAccounts/listkeys/action` and `Microsoft.Storage/storageAccounts/regeneratekey/action` whenever you want to confine the resource to a single stack.  Or you need to be really familiar with the `excludedPrincipals` on the stack.

**Comment by slavizh** [comment]
I hope that we will not have to define a big list of those actions when we create/update deployment stack. If we have to do that I would opt-in to having some default resources just like built-in roles that define which actions should be allowed.

**Comment by J0F3** [comment]
The ability to exclude  some actions would be definitely meaningful and important. But my concern is also that the list may get very long and difficult to manage. 
I would therefore also welcome a possibility to have some “defaults” which can be specified. I like the idea to just specify a built-in role an then all the allowed action from it would be excluded. But I do not know how hard that would be to implement though.

**Comment by loekd** [comment]
Could you elaborate on the reasoning behind the choice between `denyAssignment` (1) and storing state about managed resource Id's within a deployment stack (2)? 
Seeing all the downsides of the first, I wonder if the second isn't easier to implement.

**Comment by bmoore-msft** [comment]
The first `denyAssignment` is far easier to implement - given we already have that capability in the design, it leverages the platforms RBAC capabilities.  Implementing something "new" may provide a better experience but is more work and has little margin for error in a new platform capability.

**Comment by aperezmicrosoft** [comment]
Hi Everyone, as of now we are not planning on taking on this work given the technical complexity of it. If you feel this design decision will prevent you from leveraging Deployment Stacks, please let us know.

**Comment by miqm** [comment]
Just to clarify - it'll still be possible for one stack to, manage i.e. Sql server and second stack to manage database within it?

**Comment by alex-frankel** [comment]
Yes

**Comment by loekd** [comment]
This would have been a great feature, but if it requires modifications to the Azure core I can understand it's not that easy. 

It's not a blocker I suppose, but I do believe this will cause issues for end-users in more complex environments. (accidentally conflicting lifecycles, locks, deny assignments)

**Comment by aperezmicrosoft** [comment]
Adding work item reference: https://msazure.visualstudio.com/One/_backlogs/backlog/Azure-ARM-Deployments/Features/?workitem=13227840
